dubbo-remoting/dubbo-remoting-netty4/src/main/java/org/apache/dubbo/remoting/transport/netty4/NettyClientHandler.java

```diff
@@ -78,27 +76,33 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
         }
     }
 
-
     @Override
     public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
         super.write(ctx, msg, promise);
-        NettyChannel channel = NettyChannel.getOrAddChannel(ctx.channel(), url, handler);
-        try {
-            // if error happens from write, mock a BAD_REQUEST response so that invoker can return immediately without
-            // waiting until timeout. FIXME: not sure if this is the right approach, but exceptionCaught doesn't work
-            // as expected.
-            if (promise.cause() != null && msg instanceof Request) {
+        final NettyChannel channel = NettyChannel.getOrAddChannel(ctx.channel(), url, handler);
+        final boolean isRequest = msg instanceof Request;
+
+        // We add listeners to make sure our out bound event is correct.
+        // If our out bound event has an error (in most cases the encoder fails),
+        // we need to have the request return directly instead of blocking the invoke process.
+        promise.addListener(future -> {
+            if (future.isSuccess()) {
+                // if our future is success, mark the future to sent.
+                try {
+                    handler.sent(channel, msg);
+                } finally {
+                    NettyChannel.removeChannelIfDisconnected(ctx.channel());
+                }
```

Comment:
The `finally` scope has shrank compared to before.

I think it should also cover 

```java
handler.received(channel, response);
```

Reply:
That's my only concern with this PR.

Reply:
My bad.
I will fix it..